paragraph/bridging/hlo/hlo_converter.cc

```diff
@@ -0,0 +1,300 @@
+           : computation->MakeInstructionPostOrder()) {
+    TF_ASSIGN_OR_RETURN(paragraph::InstructionProto instr_proto,
+                        HloInstructionToInstructionProto(instr,
+                                                         cost_analysis,
+                                                         new_id));
+    // If new_id == 0 then we keep ids from HLO instruction, otherwise we create
+    // new unique ids for each instruction
+    int64_t old_instr_id = instr->unique_id();
+    if (new_id == 0) {
+      instr_proto.set_instruction_id(old_instr_id);
+    } else {
+      TF_RET_CHECK(instruction_ids_map.find(old_instr_id) !=
+                   instruction_ids_map.end());
+      instr_proto.set_instruction_id(instruction_ids_map.at(old_instr_id));
+    }
+    // Check root id for current computation and set it if root is found
+    if (old_instr_id == computation->root_instruction()->unique_id()) {
+      proto.set_subroutine_root_id(instr_proto.instruction_id());
+    }
+    // Connect instruction with its operands
+    for (const xla::HloInstruction* operand : instr->operands()) {
+      TF_RET_CHECK(instruction_ids_map.find(operand->unique_id()) !=
+                   instruction_ids_map.end());
+      instr_proto.add_operand_ids(instruction_ids_map.at(operand->unique_id()));
+    }
+    proto.add_instructions()->Swap(&instr_proto);
+  }
+  // Check we found root for the subroutine
+  TF_RET_CHECK(proto.subroutine_root_id() != 0);
+  return proto;
+}
+
+xla::StatusOr<paragraph::GraphProto> HloModuleToGraphProto(
+    const xla::HloModule* module,
+    const ComputeCostAnalysis* cost_analysis,
+    bool reset_ids) {
+  paragraph::GraphProto proto;
+  proto.set_name(module->name());
+  proto.set_processor_id(-1);
+
+  // If next_unique_id set to 0, we would pick the old ids from Hlo module.
+  // Setting it to 1 we reset all ids in the Hlo
+  int64_t next_unique_id = 0;
+  if (reset_ids) {
+    next_unique_id = 1;
+  }
+
+  TF_ASSIGN_OR_RETURN(
+      auto entry_subroutine_proto,
+      HloComputationToSubroutineProto(module->entry_computation(),
+                                      cost_analysis,
+                                      next_unique_id));
+  entry_subroutine_proto.set_execution_probability(1.0);
+  entry_subroutine_proto.set_execution_count(1);
+  *proto.mutable_entry_subroutine() = entry_subroutine_proto;
+  return proto;
+}
+
+xla::StatusOr<std::unique_ptr<ComputeCostAnalysis>> CreateAndRunCostAnalysis(
+        const xla::HloModule* module,
+        const ComputeCostAnalysis::Properties per_second_rates) {
+  auto cost_analysis = absl::make_unique<ComputeCostAnalysis>(
+      ShapeSize, per_second_rates);
+  xla::HloComputation* hlo_computation = module->entry_computation();
+  TF_CHECK_OK(hlo_computation->Accept(cost_analysis.get()));
+  TF_CHECK_OK(cost_analysis->UpdateInstructionProperties());
+  return cost_analysis;
+}
+
+xla::StatusOr<paragraph::GraphProto> HloConverter(
```

Comment:
This is the only public function declared in the header file. Put all others in an anonymous namespace.

Reply:
Done